### PR TITLE
drivers: spi: stm32h7: Use SPI FIFO

### PR DESCRIPTION
**Pull request by dgastonochoa**
This is a proposal to support different "packet sizes" in the SPI driver. In this context, a packet size bigger than one allows to write more than one frame to the SPI's TxFIFO (in case it has one) so that delays can be avoided:

 * It allows to remove delays between frames.
 * If interrupts are enabled, the same data will be sent using less interrupts, as each one will send several frames.

To ilustrate the first point, here are some captures of a transmission with the stm32h7 SPI current driver:

with_gaps

As you can see above, there are delays between each frame. This can dramatically impair performance when sending large amounts of data.

If a bigger packet size is used for the same transmission, the delays disappear:

without_gaps

EDIT
------
After the received feedback, this PR will no longer use the FIFO threshold. It now implements the use of the FIFO but without using modifying the FIFO threshold (so it defaults to 1). This still fixes the performance problem described above.

Test plan
=========

Connect an STM32H7 board to the PC with pins D11 and D12 connected.

Open a minicom or equivalent terminal to read the test report.

Execute the following commands:

```
rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_test_packet_sizes.loopback
west flash
```

And verify all tests pass.

EDIT
-------
The tests above are no longer necessary as the use of the FIFO threshold has been discarded. Thus, run the already existing spi loopback tests and verify they pass:

```
rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi tests/drivers/spi/spi_loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_16bits_frames.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_dma.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_16bits_frames_dma.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_dma_no_nocache.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_16bits_frames_dma_no_nocache.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_interrupt.loopback
west flash




rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h743zi tests/drivers/spi/spi_loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h743zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_16bits_frames.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h743zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_dma.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h743zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_16bits_frames_dma.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h743zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_dma_no_nocache.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h743zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_16bits_frames_dma_no_nocache.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_interrupt.loopback
west flash
```

EDIT 2
---------
This PR has also been tested with 2 separated nucleo boards (H743), one acting as SPI master and the other as SPI slave.

### CODE REVIEW

File: drivers/spi/spi_ll_stm32.c [hunk 1]
```diff
@@ -326,6 +326,39 @@ static void spi_stm32_read_next_frame(SPI_TypeDef *spi,
 	}
 }
 
+#if SPI_STM32_FIFO_ENABLED
+static void spi_stm32_send_next_packet(SPI_TypeDef *spi,
+		struct spi_stm32_data *data)
+{
+	uint8_t packet_size = SPI_PACKET_SIZE_GET(data->ctx.config->operation);
```
Comment (tbursztyka) [inline drivers/spi/spi_ll_stm32.c:333]: Packet is really not relevant in SPI. Packet holds a meaning about a start and an end of data (thus its usage in network for instance)
In SPI, having a fifo in the controller does not mean you are going to send/receive "packets". 

You are just feeding/reading a fifo according to a threshold. Imo, it would more relevant to use that word.
You would just need to loop on sending/reading around data->threshold

File: drivers/spi/spi_ll_stm32.c [hunk 2]
```diff
@@ -326,6 +326,39 @@ static void spi_stm32_read_next_frame(SPI_TypeDef *spi,
...
+
+	for (int i = 0; i < packet_size; i++) {
+		spi_stm32_send_next_frame(spi, data);
```
Comment (tbursztyka) [inline drivers/spi/spi_ll_stm32.c:336]: I don't know how stm32's SPI controllers work, generally fifo interrupt threshold is meant to make sure that the controller is not going to starve of frames to send (if there are way more to be sent, that is). Which means that between the time you get the interrupt and handle it, the controller may have had time to continue sending frames left in the fifo, which in return, means that it went over the threshold so you'll need to feed more frames than the configured threshold. It can also be true in receiving mode in a RX only transaction type where the controller can send the dummy bytes by itself (I don't kmow if that applies to this controller though).

**Review by tbursztyka — DISMISSED** [review]
fifo handling needs more thoughts I think

File: drivers/spi/spi_ll_stm32.c [hunk 3]
```diff
@@ -364,6 +400,7 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 	}
 
 	spi_stm32_read_next_frame(spi, data);
+#endif
```
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:403]: Because that would modify logic that will be used for other SPI controllers apart from the H7's. I don't have access to these platforms, so I can't test these changes on them; I want to only modify logic I can test, as much as possible.

File: drivers/spi/spi_ll_stm32.c (hunk 1 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:333]: Given that this is a driver for stm32, I opted for using the nomenclature that stm32 already uses, so potential users understand it better as the code refers to the same concepts as the stm32 datasheet and HAL library. 

E.g. see the datasheet capture below relating the FIFO threshold with packet size:

ifoth


Also, see the following function generated by STM32CubeMX:

```
/**
  * @brief  Compute configured packet size from fifo perspective.
  * @param  hspi: pointer to a SPI_HandleTypeDef structure that contains
  *               the configuration information for SPI module.
  * @retval Packet size occupied in the fifo
  */
static uint32_t SPI_GetPacketSize(SPI_HandleTypeDef *hspi)
{
  uint32_t fifo_threashold = (hspi->Init.FifoThreshold >> SPI_CFG1_FTHLV_Pos) + 1UL;
  uint32_t data_size       = (hspi->Init.DataSize      >> SPI_CFG1_DSIZE_Pos) + 1UL;

  /* Convert data size to Byte */
  data_size = (data_size + 7UL) / 8UL;

  return data_size * fifo_threashold;
}

```

This argument is applicable only to the stm32 SPI driver, not to the `spi.h` Zephyr interface.

File: drivers/spi/spi_ll_stm32.c (hunk 2 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:336]: > I don't know how stm32's SPI controllers work, generally fifo interrupt threshold is meant to make sure that the controller is not going to starve of frames to send (if there are way more to be sent, that is).

That's my understanding as well, yes.

> Which means that between the time you get the interrupt and handle it, the controller may have had time to continue sending frames left in the fifo, which in return, means that it went over the threshold so you'll need to feed more frames than the configured threshold. 

I understand you are suggesting that, instead of writing frames to the TxFIFO until the packet size is reached, this should keep writing (or reading) frames as long as TXP (which indicates that the FIFO has room for at least one packet more) is set. If that's the case, yes, it could be done that way.

File: include/zephyr/drivers/spi.h [hunk 4]
```diff
@@ -138,6 +138,28 @@ extern "C" {
...
+
+#define SPI_PACKET_SIZE_GET(_operation_)                                        \
+	((((_operation_) & SPI_PACKET_SIZE_MASK) >> SPI_PACKET_SIZE_SHIFT)          \
+	    + SPI_PACKET_SIZE_OFFSET)
+
+#define SPI_PACKET_SIZE_SET(_packet_size_)                                      \
+		(((_packet_size_) - SPI_PACKET_SIZE_OFFSET) <<                          \
+			SPI_PACKET_SIZE_SHIFT)
+/** @} */
+#endif /* CONFIG_SPI_EXTENDED_MODES */
```
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: What would be better place to put this instead? The `spi_ll_stm32.h` header is private. What is the standard Zephyr approach for platform-specific definitions in a driver?

File: drivers/spi/spi_ll_stm32.c [hunk 5]
```diff
@@ -272,6 +272,60 @@ static int spi_dma_move_buffers(const struct device *dev, size_t len)
 /* Value to shift out when no application data needs transmitting. */
 #define SPI_STM32_TX_NOP 0x00
 
+static void spi_stm32_send_next_frame(SPI_TypeDef *spi,
+		struct spi_stm32_data *data)
+{
+	const uint8_t frame_size = SPI_WORD_SIZE_GET(data->ctx.config->operation);
```
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:278]: It would be cleaner and easier to retrieve. However, the same information would be replicated in two places, with the inconsistency problems that could carry. I don't have a strong opinion on this so I'm happy to add a new attribute to the struct.

File: drivers/spi/spi_ll_stm32.c [hunk 6]
```diff
@@ -272,6 +272,60 @@ static int spi_dma_move_buffers(const struct device *dev, size_t len)
...
+		spi_context_update_tx(&data->ctx, 1, 1);
+		break;
+	case 16:
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			tx_frame = UNALIGNED_GET((uint16_t *)(data->ctx.tx_buf));
+		}
+		LL_SPI_TransmitData16(spi, tx_frame);
+		spi_context_update_tx(&data->ctx, 2, 1);
+		break;
+	default:
```
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:296]: > you don't need that. Once you configured your controller, frame size will not change.

You mean the `switch` statement? I replaced the `if ... else ...` by `switch` because it leaves the code in a better position to support more frame sizes in the future (H7 has quite a lot of them). However, I don't mind reverting this back to `if ... else ...`.

File: include/zephyr/drivers/spi.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: About the depth of the FIFO, it would be good to add a new parameter to indicate this to `st,stm32-spi-fifo.yaml`, so the driver can return an error if the FIFO threshold (or packet size as it stands now in this PR) set by the user is too big. I'm happy to do this within this PR if you agree.

[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: I have taken a quick look at the current Zephyr's drivers to find out more about this. For instance, the ADC driver has a directory `include/zephyr/drivers/adc` which contains (among others) the 2 following header files:

 * lmp90xxx.h
 * ads114s0x.h

These seem to be device-specific headers. Maybe I could apply the same principle to this PR, that is, to move the `SPI_PACKET_SIZE` definitions to a file like `include/zephyr/drivers/spi/stm32.h`.

File: drivers/spi/spi_ll_stm32.c [hunk 7]
```diff
@@ -326,6 +326,39 @@ static void spi_stm32_read_next_frame(SPI_TypeDef *spi,
 	}
 }
 
+#if SPI_STM32_FIFO_ENABLED
+static void spi_stm32_send_next_packet(SPI_TypeDef *spi,
+		struct spi_stm32_data *data)
```
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:331]: Agreed

**Comment by dgastonochoa** [comment]
Kind reminder to review this @tbursztyka. I agree with several of your comments, but others require a further discussion.

**Comment by MaureenHelm** [comment]
> Kind reminder to review this @tbursztyka. I agree with several of your comments, but others require a further discussion.

@dgastonochoa @tbursztyka @teburd can you attend the [dev meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings) this week?

File: include/zephyr/drivers/spi.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/zephyr/drivers/spi.h:161]: Headers that end up in include/zephyr/drivers/<api>/ are about drivers that let some features or configuration options to be manipulated by the user program.

In your case, you don't need to expose anything publicly. Actually, adding a new parameter to the dts binding as you noticed, would be the way to go.

File: drivers/spi/spi_ll_stm32.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply drivers/spi/spi_ll_stm32.c:296]: Frame size can be of below a byte or above, but less than 2 bytes etc... (I.E.: 4, 6, ... 10, 12...). But in the end you handle everything on byte-basis. So even if it's below a byte: you feed one byte (the controller will only take the relevant n bits representing the frame). Above a byte but below 2bytes: you feed 2 bytes. etc...

Stm32's HAL does not seem to provide anything but LL_SPI_TransmitData\<8/16/32\>. 

if you enable sizes above 16bits, then perhaps keep the switch. if/else if/else is less clean than a switch in such case.

File: drivers/spi/spi_ll_stm32.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply drivers/spi/spi_ll_stm32.c:403]: As you want, it clutters the code a bit though.

File: drivers/spi/spi_ll_stm32.c (hunk 1 above)
Comment (tbursztyka) [reply drivers/spi/spi_ll_stm32.c:333]: weird naming choice from ST then :)

File: drivers/spi/spi_ll_stm32.c (hunk 2 above)
Comment (tbursztyka) [reply drivers/spi/spi_ll_stm32.c:336]: yep, make it a while loop. The little extra time spent doing so, is less of an issue for the system than handling more interrupts due to a threshold that could hit more often.

File: drivers/spi/spi_ll_stm32.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply drivers/spi/spi_ll_stm32.c:278]: More pointer dereference might increase the risk of cache miss. And since in this case you are only reaching for a small info, so better storing it in data right away

**Comment by dgastonochoa** [comment]
@MaureenHelm I won't be able to attend to the dev meeting today, but I will the next week. Maybe you want to consider postponing the review of this PR until then if it's more convenient for you and @tbursztyka and @teburd.

File: include/zephyr/drivers/spi.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: Notice that I referred specifically to the FIFO depth to be added to device tree. The FIFO threshold shouldn't be added to devicetree: it might vary depending on the length of the data to be sent, which is provided in the SPI configuration struct. For example, if the FIFO threshold is set as 8 bytes, a transfers of size 9 bytes will lock for ever. This is because RXP is set when at least 1 packet is available in the RxFIFO, meaning that the last byte sent won't make RXP to be set, hence preventing the SPI driver to read the last byte of the transmission.

File: drivers/spi/spi_ll_stm32.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:278]: Agreed.

File: drivers/spi/spi_ll_stm32.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:296]: Agreed, I will revert to `if else`.

**Comment by MaureenHelm** [comment]
> @MaureenHelm I won't be able to attend to the dev meeting today, but I will the next week. Maybe you want to consider postponing the review of this PR until then if it's more convenient for you, @tbursztyka and @teburd.

No problem, I'll push it out to next week.

File: include/zephyr/drivers/spi.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: @tbursztyka this was discussed in the dev meeting and it was approved that the FIFO depth should be in devicetree and the FIFO threshold/packet size should be provided along with the other spi config. parameters. Do you see any problem with this?

[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/drivers/spi.h:161]: Packet size terminology seems very specific to st here. NXP's LPSPI uses the term frame size for I believe something similar. But really shouldn't all this be done internally by the driver to optimally send whatever data is there? Maybe I missed something

[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: ST uses frame size as well, but to define how the FIFO is addressed (8, 16, 32 bits). Packet size is given in units of frames. Thus, a packet size of 3 is 3 bytes if the frame size is 8 and it's 12 bytes if frame size is 32 bits. In any case, I'm happy to name this FIFO threshold instead.

It's true that an 'optimum' (biggest possible) FIFO threshold can be inferred from the FIFO depth (defined in DT) and the transfer total length, but this has 2 drawbacks:
 * It doesn't give the user the option to choose the desired FIFO threshold. It could happen that the biggest possible FIFO threshold doesn't work because the target SPI device cannot cope with it.
 * Testing becomes more confusing and difficult. It's very important to test this feature intensively as it can cause the SPI peripheral to block for ever if misused. I want to remark that test cases become way more difficult to understand. 

I had that implementation before and decided to change it to this.

[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: Kind reminder to @tbursztyka and @teburd to clarify this point please. After all the explanation above, there are 2 options:
 * Provide the FIFO threshold as a parameter
 * Calculate the optimal FIFO threshold inside the driver automatically.

I strongly suggest the first option, putting the required macros in a H7 SPI specific header file as explained in another comment. Please could you confirm if you agree?

[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/drivers/spi.h:161]: I guess the question is how does this play out in an independent driver for a SPI device, like a sensor driver going to read a 2kb FIFO over SPI.

In this case, to get the ideal performance it seems like for this particular platform some ifdefry would be needed. That to me seems like a good indicator that exposing these details is the wrong approach but I could be wrong.

[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h:161]: > I guess the question is how does this play out in an independent driver for a SPI device, like a sensor driver going to read a 2kb FIFO over SPI.

With the current implementation, the FIFO threshold defaults to 1, so other modules using an SPI driver won't notice any change. Possibly they will still notice an improvement in performance.

> In this case, to get the ideal performance it seems like for this particular platform some ifdefry would be needed. 

Yes, a module using this SPI driver would need to check (potentially with an #ifdef) if it is an H7 and then provide a specific FIFO threshold for the particular length it's trying to send.

If this is a problem, I believe (I would need to test it though) that it would be possible to not provide a FIFO threshold in any way or form. This would mean that H7 would always work with FIFO threshold = 1, but it would still use the FIFO capabilities, at least after applying the feedback below, where data is written/read to/from the FIFO while TXP/RXP allows it.

[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h]: I have removed the use of packet size/FIFO threshold.

File: drivers/spi/spi_ll_stm32.c (hunk 1 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c]: Solved as this name is no longer used.

File: drivers/spi/spi_ll_stm32.c (hunk 2 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c]: Using a while loop here caused several problems produced by the fact that SPI supports transmissions in which tx is bigger than rx or vice-versa. The current implementation takes care of this, defaulting to the classic way of transmitting in case of rx > tx (which causes more problems that require further modifications).

File: drivers/spi/spi_ll_stm32.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c]: Done

File: drivers/spi/spi_ll_stm32.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:278]: I have taken a look at this. There are other places in which `ctx.config` is accessed; to be consistent, all these references should be modified. This is out of this PR's scope as I conceived it.

File: drivers/spi/spi_context.h [hunk 8]
```diff
@@ -287,6 +289,21 @@ static inline void *spi_context_get_next_buf(const struct spi_buf **current,
 	return NULL;
 }
 
+static inline size_t spi_stm32_count_total_bytes(const struct spi_buf_set *bufs)
```
Comment (erwango) [inline drivers/spi/spi_context.h:292]:
```suggestion
static inline size_t spi_count_total_bytes(const struct spi_buf_set *bufs)
```
And this modification should be done in a dedicated commit with proper justification

File: drivers/spi/spi_ll_stm32.c [hunk 9]
```diff
@@ -338,20 +338,57 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
...
 }
 
-/* Shift a SPI frame as master. */
-static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+static int spi_stm32_can_use_fifo(SPI_TypeDef *spi,
+								  struct spi_stm32_data *data)
+{
+	ARG_UNUSED(spi);
+
+#if SPI_STM32_FIFO_ENABLED
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:346]: Why not using `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)` directly ?
H7 compatible SPI IPs are variants of SPI-FIFO IPs

File: drivers/spi/spi_ll_stm32.c [hunk 10]
```diff
@@ -338,20 +338,57 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
 	return 0;
 }
 
-/* Shift a SPI frame as master. */
-static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+static int spi_stm32_can_use_fifo(SPI_TypeDef *spi,
+								  struct spi_stm32_data *data)
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:342]: Nit: indentation

**Review by erwango — CHANGES_REQUESTED** [review]
Nice contribution.
Some changes required for my part

File: drivers/spi/spi_context.h (hunk 8 above)
Comment (dgastonochoa) [reply drivers/spi/spi_context.h:292]: This modification is included in the "Use SPI FIFO" commit because it's related to it: it's necessary to decide if the SPI FIFO can be used. The other commits could be considered as related, but they are useful by themselves, unlike this one, that would be only adding an extra parameter which, if not used, only adds complexity to the code.

File: drivers/spi/spi_ll_stm32.c (hunk 9 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:346]: Normally, the actual check should be only if the SPI device in question has a FIFO (indicated in devicetree). However, I add the H7 check as well as that's the only platform I can test. I could leave the H7 check only, but I wanted to remark that the actual criteria to enable this (or not) is whether or not a FIFO is available, so I left both checks so another user can (hopefully) understand the intention.

File: drivers/spi/spi_ll_stm32.c (hunk 10 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:342]: I tried to indent this better at first: by using a tab at the beginning of the line and then all spaces until it gets aligned with the `spi` parameter. The indentation checks threw an error. 

Another solution is to add a couple of tabs instead of aligning it with the `spi` parameter. I used to do this until I was told that parameters should be aligned. I believe that comment is in this PR.

What is the actual criteria to be followed?

File: drivers/spi/spi_context.h (hunk 8 above)
Comment (erwango) [reply drivers/spi/spi_context.h:292]: First issue here is that you're modifying a generic driver file using vendor prefix. This isn't allowed.
Then it is good practice in general to split API changes from vendor specific implementations (it could be useful for instance a revert has to be done on the vendor code modification).

File: drivers/spi/spi_ll_stm32.c (hunk 10 above)
Comment (erwango) [reply drivers/spi/spi_ll_stm32.c:342]: You should use 8 chars-tabs and complete with spaces for the last 7 indentation chars.

File: drivers/spi/spi_ll_stm32.c (hunk 9 above)
Comment (erwango) [reply drivers/spi/spi_ll_stm32.c:346]: This is misleading as using `SPI_STM32_FIFO_ENABLED`, it's hard to guess that code impacts only H7 compatible devices.
Question is: Is that code compatible with all `st,stm32-spi-fifo` compatible devices ? or is that specific to `st,stm32h7-spi` ones ?

Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:346]: > Question is: Is that code compatible with all st,stm32-spi-fifo compatible devices ? or is that specific to st,stm32h7-spi ones ?

I can't answer this question because I can only test this change for H7. It _should_ be compatible with `st,stm32-spi-fifo`, but I can't assure it as I can't test it. I want to restrict the change to H7 to avoid introducing bugs or breaking applications already using Zephyr.

However, I can add a comment to explain the purpose of checking H7.

Comment (erwango) [reply drivers/spi/spi_ll_stm32.c:346]: > However, I can add a comment to explain the purpose of checking H7.

Please do the opposite:
Use a check on `st,stm32h7-spi`. Add a comment saying it has to be verified on `st,stm32-spi-fifo`.
It will be much easier to understand quickly the situation.

File: drivers/spi/spi_context.h (hunk 8 above)
Comment (dgastonochoa) [reply drivers/spi/spi_context.h]: Done.

File: drivers/spi/spi_ll_stm32.c (hunk 10 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c:342]: I believe the indentation is correct now.

File: drivers/spi/spi_ll_stm32.c (hunk 2 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c]: I reviewed this and found a clean way to support the rx > tx case, so this is no longer a problem.

File: include/zephyr/drivers/spi.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgastonochoa) [reply include/zephyr/drivers/spi.h]: @teburd @tbursztyka unless you have any objection, I will resolve this conversation.

File: drivers/spi/spi_ll_stm32.c (hunk 1 above)
Comment (dgastonochoa) [reply drivers/spi/spi_ll_stm32.c]: @teburd @tbursztyka unless you have any objection, I will resolve this conversation.

**Comment by dgastonochoa** [comment]
All feedback has been applied, @teburd and @tbursztyka please could you take a look?

**Comment by dgastonochoa** [comment]
@teburd, @tbursztyka please could you confirm whether or not you intent to review this at some point?

**Review by erwango — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Comment by GeorgeCGV** [comment]
@dgastonochoa that PR breaks some SPI devices on H7 for me (_including #57848_).
Stuck within `spi_stm32_complete` in `spi_stm32_isr` as SPI is kept busy..

**Comment by dgastonochoa** [comment]
Hi @GeorgeCGV, I have run the spi_loopback tests again with interrupts enable in with `CONFIG_DEBUG_OPTIMIZATIONS`; they pass (on H743).

I imagine you are seeing this error not in the loopback tests, but in some other scenario, right? Would it be possible to run the loopback tests on H730 to see if the problem is reproduced there? Does the problem reproduce without enabling SPI interrupts?

When you say "SPI is kept busy", do you mean that `ll_func_spi_is_busy` keeps returning true?

**Comment by GeorgeCGV** [comment]
> Does the problem reproduce without enabling SPI interrupts?

With disabled interrupts it is not getting into a freezing state. But I am observing overruns and some devices fail.

> When you say "SPI is kept busy", do you mean that ll_func_spi_is_busy keeps returning true? Would it be possible to check if spi_context_tx_on returns false (all bytes have been sent) but spi_context_rx_on returns true (not all bytes have been read)?

Yes, it gets stuck in `spi_stm32_complete` after data reception.  The `CR2` and `SR` registers are `0`. Therefore, `ll_func_spi_is_busy` returns `true`.

* `spi_context_tx_on` - the context `tx_len` is `0`.
* `spi_context_rx_on` - the context `rx_len` is `0`.

Therefore, both return `false`. The SPI devices fail not on every boot, but sometimes. Feels like it is not stable. Would need to do signal capture for details...

> Would it be possible to run the loopback tests on H730 to see if the problem is reproduced there?

Would need to do some HW modifications.

**Comment by dgastonochoa** [comment]
@GeorgeCGV 

> > Does the problem reproduce without enabling SPI interrupts?
> 
> With disabled interrupts it is not getting into a freezing state. But I am observing overruns and some devices fail.
> 
> > When you say "SPI is kept busy", do you mean that ll_func_spi_is_busy keeps returning true? Would it be possible to check if spi_context_tx_on returns false (all bytes have been sent) but spi_context_rx_on returns true (not all bytes have been read)?
> 
> Yes, it gets stuck in `spi_stm32_complete` after data reception. The `CR2` and `SR` registers are `0`. Therefore, `ll_func_spi_is_busy` returns `true`.
> 
>     * `spi_context_tx_on` - the context `tx_len` is `0`.
> 
>     * `spi_context_rx_on` - the context `rx_len` is `0`.

This seems to indicate that the SPI driver has send/received all the expected bytes, as the `tx/rx_on` functions return 0. However, it seems that either `TXC` is not set, or it is set and cleared before `spi_stm32_complete` is called, so the driver ends up in an endless loop. 

But if the only way `TXC` would remain 0 is that there is still activity in the bus. Would it be possible to confirm the activity on the bus when the error occurs?

**Comment by GeorgeCGV** [comment]
> Also, is it possible to double check that is TXC the flag that's being checked? (ll_func_spi_is_busy can check TXC, EOT or BSY depending on circumstances)

```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
	if (LL_SPI_GetTransferSize(spi) == 0) {
		return LL_SPI_IsActiveFlag_TXC(spi) == 0;
	} else {
		return LL_SPI_IsActiveFlag_EOT(spi) == 0;
	}
...
```

it is H7, so we go and check the transfer size. The `CR2` is `0`, obviously the `SPI_CR2_TSIZE` is `0` as well.
Now we go into `LL_SPI_IsActiveFlag_TXC`, `SR` is `0x2` in one case or `0x0` in the other. The `TXC` bit is `0`.
Because `TXC` is `0`, the `ll_func_spi_is_busy` returns `true`.

Different SPI peripherals fail at different times. To trigger the issue I have to power cycle the board multiple times. When it gets stuck it is always within `spi_stm32_complete` in spi isr. The `TSIZE`, `TXC`, and context rx/tx count, len are `0` as well as rx/tx buffers are set to `0x0`.

What is also observed is the corruption of transmitted and/or received data. That doesn't happen every time... but often enough.

> Would it be possible to confirm the activity on the bus when the error occurs?

Nothing happens on the bus when it stucks:

stuck_state

**Comment by dgastonochoa** [comment]
In that picture you sent there is a lot of cross-talk to CS, is it sure that that is caused bu the logic analyzer probes? Could you try to  reduce the SPI SCK frequency?

**Comment by GeorgeCGV** [comment]
@dgastonochoa it is a zoomed out screen. The clocks are validated and set to be within supported range. It is a custom device based on H730. Multiple SPIs are used where each has multiple targets (_slaves_).

After reducing the clock the issue started to appear every time. Sized down connected zoo to one device per SPI (_to eliminate possible issue(s) from having multiple devices_). 

SPI without FIFO, the device works:
fifo_off_works

SPI with FIFO, the device doesn't work and it is stuck as described above (_within `ll_func_spi_is_busy`, `SR=0`_):
fifo_on_fail

It is clearly visible that the CS (_chip select_) is dramatically off..., It is software controlled by the driver. Most probably that is where the problems are coming from.

**Comment by avisconti** [comment]
@dgastonochoa I opened an issue (#66486) that looks different from #66326, but I'm not sure. @erwango just closed it, but I'm kindly ask you to take a look to it and see if it might be the same. Thanks.
